pyteal/ast/subroutine.py

```diff
@@ -105,9 +107,7 @@ def _validate_parameter_type(
         else:
             if not isclass(ptype) and not SubroutineDefinition.is_abi_annotation(ptype):
```

Comment:
@ahangsu For my edification:  Do you understand why the following expression evaluates to `False`?

`print(isclass(pt.abi.StaticArray(pt.abi.ByteTypeSpec(), 10)))` 

* It's _slightly_ tangential to the PR.  I reviewed here while inspecting tests.  It took me some time to realize _some_ ABI arguments aren't treated as classes even though they're defined as classes.
* When I run a debugger locally, the debugger does _not_ let me step into `inspect.isinstance`.  So, I have limited visibility.
* The inline docs to `inspect.isclass` suggest classes provide `__doc__` and `__module__`.  The example _does_ provide these attributes.

Reply:
The purpose of `_validate_parameter_type` is to check the annotation in python function definition.

`pt.abi.StaticArray(pt.abi.ByteTypeSpec(), 10))` is an instance not class.

`isclass` here is to allow for `Expr` or `ScratchVar` annotations.

some other type annotation on ABI side, e.g., `abi.StaticArray[abi.Uint64, Literal[10]]`, are generic alias. This is fed to `SubroutineDefinition.is_abi_annotation`.

Reply:
Do you have an example of ABI args not treated as classes?

Reply:
Sorry, I confused things with the (incorrect) example.

I began looking here because I deleted ` and not SubroutineDefinition.is_abi_annotation(ptype)` from the highlighted line to see what tests fail.

You _should_ see 3 test failures.  All failures involve a reference to `StaticArray` like the example failure message below.

```
E               pyteal.TealInputError: Function has parameter b of declared type pyteal.StaticArray[pyteal.Uint32, typing.Literal[10]] which is not a class
```

I opened the thread because it's unclear to me why `isclass` evaluation fails here.

Reply:
let me illustrate more here... `StaticArray[abi.Byte, Literal[10]]` is *not* a class, it is a generic alias: https://docs.python.org/3/library/stdtypes.html#types-genericalias.

If you lift the condition `and not SubroutineDefinition.is_abi_annotation(ptype)`, it will be not class, and directly raise error.

Reply:
The function name itself might be somewhat misleading: `_validate_parameter_type` should be renamed as `_validate_annotation_type` I guess?

Reply:
I see what you're saying. The line says `if not isclass(ptype) and not SubroutineDefinition.is_abi_annotation(ptype)` which to me feels incorrect. I would expect all `abi_annotation`'s to satisfy `isclass()` and therefore the second statement should be tautologically true.

Reply:
I only saw @ahangsu 's explanation after putting my question forth. I think I understand now.

As far as naming, maybe just `_validate_annotation()` ?

Final, question: Does it make sense for this to be a `@staticmtehod`, or would it simplify things to just make an in instance method and access `user_defined_annotations` from `self` ?

Reply:
Yes and yes I think, I will rename this to be `_validate_annotation`, and probably I should make this a simple instance method, so we can simplify some function arguments.

Reply:
I am also reconstruct the function a little bit on my local side, so the readability would be better.

Reply:
@ahangsu and @tzaffi thanks for your efforts to respond to my question. 

* I now understand the situation.  I misunderstood that `StaticArray` takes on `GenericAlias` at runtime.  
* I briefly paired with @tzaffi to deepen my understanding here and ended up with _optional_ changes https://github.com/algorand/pyteal/pull/312.  @ahangsu I'd appreciate a review as time allows.  And feel welcomed to keep as is.